driver-core/src/main/java/com/datastax/driver/core/StatementWrapper.java

```diff
@@ -0,0 +1,152 @@
+/*
+ *      Copyright (C) 2012-2014 DataStax Inc.
+ *
+ *   Licensed under the Apache License, Version 2.0 (the "License");
+ *   you may not use this file except in compliance with the License.
+ *   You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *   Unless required by applicable law or agreed to in writing, software
+ *   distributed under the License is distributed on an "AS IS" BASIS,
+ *   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *   See the License for the specific language governing permissions and
+ *   limitations under the License.
+ */
+package com.datastax.driver.core;
+
+import java.nio.ByteBuffer;
+
+import com.datastax.driver.core.policies.LoadBalancingPolicy;
+import com.datastax.driver.core.policies.RetryPolicy;
+import com.datastax.driver.core.policies.SpeculativeExecutionPolicy;
+
+/**
+ * Base class for custom {@link Statement} implementations that wrap another statement.
+ * <p>
+ * This is intended for use with a custom {@link RetryPolicy}, {@link LoadBalancingPolicy} or
+ * {@link SpeculativeExecutionPolicy}. The client code can wrap a statement to "mark" it, or
+ * add information that will lead to special handling in the policy.
+ * <p>
+ * Example:
+ * <pre>
+ * {@code
+ * // Define your own subclass
+ * public class MyCustomStatement extends StatementWrapper {
+ *     public MyCustomStatement(Statement wrapped) {
+ *         super(wrapped);
+ *     }
+ * }
+ *
+ * // In your load balancing policy, add a special case for that new type
+ * public class MyLoadBalancingPolicy implements LoadBalancingPolicy {
+ *     public Iterator<Host> newQueryPlan(String loggedKeyspace, Statement statement) {
+ *         if (statement instanceof MyCustomStatement) {
+ *             // return specially crafted plan
+ *         } else {
+ *             // return default plan
+ *         }
+ *     }
+ * }
+ *
+ * // The client wraps whenever it wants to trigger the special plan
+ * Statement s = new SimpleStatement("...");
+ * session.execute(s);                         // will use default plan
+ * session.execute(new MyCustomStatement(s));  // will use special plan
+ * }
+ * </pre>
+ */
+public abstract class StatementWrapper extends Statement {
```

Comment:
This new class will cause a problem with QueryLogger. QueryLogger only handles the 3 pre-defined types of Statement: Regular, Bound and Batch. You should at least add an if-block for StatementWrapper instances in `QueryLogger#append(Statement, StringBuilder, int)` and "unwrap" the statement to log. 
As a side note, IMHO the method `getQueryString()`, which is currently defined twice in PreparedStatement and RegularStatement, should be pulled up into Statement class and be part of its API. Implementations would be trivial for most subclasses, except maybe BuiltStatement.